### PR TITLE
[SDPA] Fix 32-bit offset overflow in mem-efficient attention forward with attn_bias

### PR DESCRIPTION
**Pull request by bin123apple**
## Summary

`F.scaled_dot_product_attention` with the **EFFICIENT_ATTENTION** backend and an
explicit `attn_mask` **silently returns garbage** for every query row past
`2^32 / seq_len` once `num_queries * num_keys > 2^32` (i.e. S > 65536 for square
self-attention). No error, no warning, no NaN — the corrupted values look
plausible, which makes this extremely hard to attribute downstream (we found it
as silently diverging RL training metrics).

## Root cause

In `kernel_forward.h`, `query_start` is `uint32_t`:

https://github.com/pytorch/pytorch/blob/545c55c1724bafed6772a0b85fecfd09f9aa0dbb/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h#L635

and the per-row offset into the bias tensor is computed as
`query_start * p.bias_strideM` where `bias_strideM` is `int32_t`, so the multiply
happens in 32-bit unsigned arithmetic and wraps modulo 2^32:

https://github.com/pytorch/pytorch/blob/545c55c1724bafed6772a0b85fecfd09f9aa0dbb/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h#L811

Past `row = 2^32 / seq_len` the kernel reads the mask from wrapped addresses
(typically an earlier row of the same tensor), producing wrong attention for
those rows. Corruption onset at exactly `2^32 / S` (not `2^31 / S`) matches the
unsigned arithmetic.

The dropout RNG `skipahead` offset in the same kernel has the same 32-bit
multiply (`(query_start + thread_i) * p.num_keys_absolute`). The backward kernel
computes the equivalent offset with an `int64_t` `query_start`
(`kernel_backward.h`), so with dropout + large seq_len the forward and backward
dropout masks diverge past the same boundary, corrupting gradients.

The batch/head-level strides (`bias_strideB`/`bias_strideH`) were already
widened to `int64_t` previously; the per-row offset was missed. Note these
kernels were vendored from xformers, which has since removed its in-tree CUTLASS
FMHA sources, so the fix needs to land here.

## Fix

Compute both offsets in 64-bit, matching `kernel_backward.h`. Two one-line
casts; in-bounds behavior is bit-identical.

## Reproduction (any sm80+ GPU, bf16)

```python
import torch
import torch.nn.functional as F
from torch.nn.attention import SDPBackend, sdpa_kernel

S, HD = 73728, 64                        # S^2 = 1.27 * 2^32
torch.manual_seed(0)
q = torch.randn(1, 1, S, HD, device="cuda", dtype=torch.bfloat16) * 0.1
k = torch.randn(1, 1, S, HD, device="cuda", dtype=torch.bfloat16) * 0.1
v = torch.randn(1, 1, S, HD, device="cuda", dtype=torch.bfloat16)
idx = torch.arange(S, device="cuda")
causal = (idx[:, None] >= idx[None, :])[None, None]

ref = F.scaled_dot_product_attention(q, k, v, is_causal=True)   # ground truth
with sdpa_kernel(SDPBackend.EFFICIENT_ATTENTION):
    out = F.scaled_dot_product_attention(q, k, v, attn_mask=causal)

B = 2**32 // S   # predicted corruption onset = row 58254
print((out - ref)[0, 0, :B].abs().max())   # ~6e-5  (bf16 noise)
print((out - ref)[0, 0, B:].abs().max())   # ~0.39  (garbage)
```

Observed on `2.14.0.dev20260612+cu130` (H200, sm90) — corruption onset is
exactly `2^32 // S` for every S tested (73728→58254, 81920→52428, 98304→43690,
90000→47721). The MATH backend with the same mask tensor is correct everywhere,
confirming the mask contents are fine and the bug is in the kernel's indexing.
`is_causal=True` (no mask tensor) is unaffected.

## Test plan

- Added `test_mem_eff_attention_large_seq_len_attn_mask_index_overflow`
  (gated by `@largeTensorTest("12GB", "cuda")`): explicit additive causal mask
  vs `is_causal=True` on the same backend at seq_len=70000. Fails before this
  fix (max abs diff 0.54 past row 61356, exact 0 before it), passes after.
- Verified the kernel change directly on H200: compiled `kernel_forward.h`
  (pristine vs patched, same CUTLASS pin) into a standalone extension driving
  `AttentionKernel<bf16, Sm80, true, 64, 64, 64>` with identical S=73728
  inputs. Pristine reproduces the corruption bit-for-bit with the stock binary
  (max diff 0.39 past row 58254); patched returns bf16-noise-level diff
  (6.1e-5) everywhere, and is bit-identical to pristine for all in-bounds rows.

AI assistance (Claude) was used to locate the overflow and prepare this PR; the
change and verification were reviewed end-to-end by the submitter.

### CODE REVIEW

**Comment by Skylion007** [comment]
If you can reduce the test case slightly per copilot's suggestion looks good.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
This isn't also an issue in the backward pass, right?

**Comment by bin123apple** [comment]
> This isn't also an issue in the backward pass, right?

I think the backward pass is not affected. In `kernel_backward.h` the equivalent
per-row bias offset is already computed with an int64_t `query_start`, so it
doesn't overflow; this PR just brings the forward kernel in line with that.

**Review by Skylion007 — APPROVED** [review]

**Comment by bin123apple** [comment]
@Skylion007 The merge attempts have each been blocked by different unrelated ROCm flaky failures. Anything needed from my side? Thanks!
